src/gateway/environments/jsdom-sixteen/jsdom-sixteen-environment.js

```diff
@@ -0,0 +1,127 @@
+// @flow
+import type {ResourceLoader} from "jsdom";
+import type {IJSDOMSixteenConfiguration} from "./index.js";
+import type {IRenderEnvironment, RenderAPI, RenderResult} from "../../types.js";
+
+/**
+ * A render environment built to support the JSDOM 16.x API.
+ */
+export class JSDOMSixteenEnvironment implements IRenderEnvironment {
+    _configuration: IJSDOMSixteenConfiguration;
+
+    /**
+     * Create a new instance of this environment.
+     *
+     * @param {IJSDOMSixteenConfiguration} configuration
+     * Configuration for the environment.
+     */
+    constructor(configuration: IJSDOMSixteenConfiguration) {
+        if (configuration == null) {
+            throw new Error("Must provide environment configuration");
+        }
+        this._configuration = configuration;
+    }
+
+    _retrieveTargetFiles = async (
+        url: string,
+        renderAPI: RenderAPI,
+        resourceLoader: ResourceLoader,
+    ): Promise<Array<string>> => {
+        const traceSession = renderAPI.trace("Retrieving target files");
+        try {
+            /**
+             * First, we need to know what files to execute so that we can produce
+             * a render result, and we need a resource loader so that we can
+             * retrieve those files as well as support retrieving additional files
+             * within our JSDOM environment.
+             */
+            const files = await this._configuration.getFileList(url, renderAPI);
+
+            /**
+             * Now let's use the resource loader to get the files.
+             * We ignore the `FetchOptions` param of resourceLoader.fetch as we
+             * have nothing to pass there.
+             */
+            return await Promise.all(
+                files.map((f) => {
+                    const fetchResult = resourceLoader.fetch(f);
```

Comment:
Unlike with the `react-render-server`, I realised that we can reuse the custom resource loader for our file downloads here, simplifying our environment configuration. It also makes this more versatile, as the custom resource loader could just be loading files from disk for all this code cares.